decepticonlp/attacks/attack.py

```diff
@@ -0,0 +1,328 @@
+            input_dict = {}
+            for k, i in enumerate(self.input_format):
+                if i != "labels":
+                    input_dict[i] = inputs[k].long().to(self.device)
+                else:
+                    labels = inputs[k].long().to(self.device)
+            # Forward pass
+            outputs = self.model(**input_dict)
+
+            # Compute Loss
+            if type(self.criterion["loss"]) == torch.nn.modules.loss.BCELoss:
+                # Binary Classification with Sigmoid
+                if self.apply_activation:
+                    outputs = torch.sigmoid(outputs)
+                computed_loss = self.criterion["loss"](outputs.float(), labels.float())
+
+            elif (
+                type(self.criterion["loss"]) == torch.nn.modules.loss.BCEWithLogitsLoss
+            ):
+                # Binary Classification, but BCEWithLogitsLoss computed sigmoid on its own, so it is assumed that the user does not have sigmoid in his model
+                # apply_activation has no significance here
+                computed_loss = self.criterion["loss"](
+                    outputs.float().view(labels.shape), labels.float()
+                )
+                outputs = torch.sigmoid(outputs)
+
+            elif type(self.criterion["loss"]) == torch.nn.modules.loss.NLLLoss:
+                # Multiclass Classification, with Log_Softmax
+                if self.apply_activation:
+                    outputs = torch.log_softmax(outputs, dim=1)
+                computed_loss = self.criterion["loss"](outputs, labels)
+
+            elif type(self.criterion["loss"]) == torch.nn.modules.loss.CrossEntropyLoss:
+                # Multiclass Classification, but since CrossEntropy computes log_softmax on its own, it is assumed that the user does not have a softmax/log_softmax in his model
+                # apply_activation has no significance here
+                computed_loss = self.criterion["loss"](outputs, labels)
+                outputs = torch.log_softmax(outputs, dim=1)
+
+            running_loss += computed_loss.item()
+            loss = running_loss / (counter * dataloader.batch_size)
+
+            # Log the loss
+            if counter % self.logs_after_every == 0:
+                loss_logs.append(loss)
+
+            tqdm_dataloader.set_description("Loss (%f)" % (loss))
+
+            # Compute Accuracy
+            if self.criterion["accuracy"]:
+
+                if (type(self.criterion["loss"]) == torch.nn.modules.loss.BCELoss) or (
+                    type(self.criterion["loss"])
+                    == torch.nn.modules.loss.BCEWithLogitsLoss
+                ):
+                    computed_labels = (outputs > self.threshold).float()
+
+                elif (
+                    type(self.criterion["loss"]) == torch.nn.modules.loss.NLLLoss
+                ) or (
+                    type(self.criterion["loss"])
+                    == torch.nn.modules.loss.CrossEntropyLoss
```

Comment:
The correct way to check these things for an if statement is to use `isinstance(self.criterion["loss"], torch.nn.modules.loss.CrossEntropyLoss)`

Reply:
Will edit.